src/query/src/range_select/plan.rs

```diff
@@ -216,30 +358,226 @@ impl ExecutionPlan for RangeSelectExec {
             schema: self.schema.clone(),
+            range_exec: self.range_exec.clone(),
             input,
+            random_state: RandomState::new(),
+            time_index,
+            align: self.align,
+            by: self.by.clone(),
+            series_map: HashMap::new(),
+            exec_state: ExecutionState::ReadingInput,
+            output_num_rows: 0,
+            row_converter,
+            modify_map: HashMap::new(),
+            metric: baseline_metric,
         }))
     }
 
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metric.clone_inner())
+    }
+
     fn statistics(&self) -> Statistics {
         self.input.statistics()
     }
 }
 
-pub struct RangeSelectStream {
+struct RangeSelectStream {
+    /// the schema of output column
     schema: SchemaRef,
+    range_exec: Vec<RangeFnExec>,
     input: SendableRecordBatchStream,
+    /// Column index of TIME INDEX column's position in schema
+    time_index: usize,
+    /// the unit of `align` is millisecond
+    align: i64,
+    by: Vec<Arc<dyn PhysicalExpr>>,
+    exec_state: ExecutionState,
+    /// Converter for the by values
+    row_converter: RowConverter,
+    random_state: RandomState,
+    /// key: time series's hash value
+    /// value: time series's state on different align_ts
+    series_map: HashMap<u64, SeriesState>,
+    /// key: (hash of by rows, align_ts)
+    /// value: [row_ids]
+    /// It is used to record the data that needs to be aggregated in each time slot during the data update process
+    modify_map: HashMap<(u64, i64), Vec<u32>>,
+    /// The number of rows of the final output
+    output_num_rows: usize,
+    metric: BaselineMetrics,
+}
+
+struct SeriesState {
+    /// by values write by `RowWriter`
+    row: OwnedRow,
+    /// key: align_ts
+    /// value: a vector, each element is a range_fn follow the order of `range_exec`
+    align_ts_accumulator: HashMap<i64, Vec<Box<dyn Accumulator>>>,
+}
+
+impl RangeSelectStream {
+    fn evaluate_many(
+        &self,
+        batch: &RecordBatch,
+        exprs: &[Arc<dyn PhysicalExpr>],
+    ) -> DfResult<Vec<ArrayRef>> {
+        exprs
+            .iter()
+            .map(|expr| {
+                let value = expr.evaluate(batch)?;
+                Ok(value.into_array(batch.num_rows()))
+            })
+            .collect::<DfResult<Vec<_>>>()
+    }
+
+    fn update_range_context(&mut self, batch: RecordBatch) -> DfResult<()> {
+        let _timer = self.metric.elapsed_compute().timer();
+        let num_rows = batch.num_rows();
+        let by_arrays = self.evaluate_many(&batch, &self.by)?;
+        let mut hashes = vec![0; num_rows];
+        create_hashes(&by_arrays, &self.random_state, &mut hashes)?;
+        let by_rows = self.row_converter.convert_columns(&by_arrays)?;
+        let ts_column = batch
+            .column(self.time_index)
+            .as_any()
+            .downcast_ref::<TimestampMillisecondArray>()
+            .unwrap();
```

Comment:
Return an error instead of panic.

Reply:
Here follow the implementation of PromQL physical plan like `src/promql/src/extension_plan/instant_manipulate.rs`, I checked times index data type in logical plan stage,  the cast can not fail